gemini/reasoning-engine/tutorial_langgraph_cloud_sql_pg_rag_agent.ipynb

```diff
@@ -0,0 +1,1078 @@
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "BF8xqd84351O"
+      },
+      "source": [
+        "### Define router\n",
+        "\n",
+        "Then, you'll define a router to control the flow of the conversation, determining which tool to use based on user input or the state of the interaction. Here we'll use a simple router setup, and you can customize the behavior of your router to handle multiple tools, custom logic, or multi-agent workflows."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 26,
+      "metadata": {
+        "id": "7m9XRSn3351O"
+      },
+      "outputs": [],
+      "source": [
+        "def router(\n",
+        "    state: List[BaseMessage],\n",
+        ") -> Literal[\"book_similarity_search\", \"movie_similarity_search\", \"__end__\"]:\n",
+        "    if not state[0].content or len(state[1].tool_calls) == 0:\n",
+        "        return \"__end__\"\n",
+        "    if \"book\" in state[0].content:\n",
+        "        return \"book_similarity_search\"\n",
+        "    if \"movie\" in state[0].content:\n",
+        "        return \"movie_similarity_search\"\n",
+        "    return \"__end__\""
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "CFeOIbed351O"
+      },
+      "source": [
+        "## Define LangGraph application\n",
+        "\n",
+        "Now you'll bring everything together to define your LangGraph application as a custom template in Reasoning Engine.\n",
+        "\n",
+        "This application will use the tool and router that you just defined. LangGraph provides a powerful way to structure these interactions and leverage the capabilities of LLMs.\n",
+        "\n",
+        "The workflow:\n",
+        "```\n",
+        "graph LR\n",
+        "    A[\"Checker\"] -- \"(user question + question type)\" --> B{Router};\n",
+        "    B -- (Movie Question) --> C[Movie RAG];\n",
+        "    B -- (Book Question) --> D[Book RAG];\n",
+        "    C[Movie RAG] -- \"Retrieved Movie Information\" --> E[Summarizer]\n",
+        "    D[Book RAG] -- \"Retrieved Book Information\" --> E[Summarizer]\n",
+        "    E --> F[Answer]\n",
+        "\n",
+        "    subgraph \"Input & Initial Processing\"\n",
+        "        A[Checker]\n",
+        "        end\n",
+        "\n",
+        "    subgraph \"Content Retrieval & Processing\"\n",
+        "        B{Router}\n",
+        "        C[Movie RAG]\n",
+        "        D[Book RAG]\n",
+        "        E[Summarizer]\n",
+        "    end\n",
+        "\n",
+        "    subgraph \"Output\"\n",
+        "        F[Answer]\n",
+        "    end\n",
+        "\n",
+        "```"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "tZYtR6-zqudb"
+      },
+      "source": [
+        "#### Multi stage"
+      ]
+    },
+    {
+      "cel
```

Comment:
Suggest rendering the output of the following test cells to show users the actual output vs. expected output.

Reply:
Added the actual outputs when running the test cells.